test/virtual/host.cpp

```diff
@@ -8,16 +8,151 @@
 #endif
 
 #include <openenclave/edger8r/host.h>
+#include <mutex>
 
 #include "enclave_impl.h"
 
+/* Temporily set value. */
+#define OE_MAX_ECALLS 256
+
 extern "C"
 {
+    const char* _ecall_table[OE_MAX_ECALLS];
+    uint32_t _ecall_table_size;
+    std::mutex oe_lock;
     thread_local oe_enclave_t* _enclave;
 
+    // Get the global ecall id from the _ecall_table.
+    static oe_result_t oe_get_global_ecall_id_by_name(
+        const char* name,
+        uint64_t* global_id)
+    {
+        oe_result_t result = OE_NOT_FOUND;
+        uint64_t i;
+
+        oe_lock.lock();
+        if (!global_id)
+        {
+            result = OE_INVALID_PARAMETER;
+            goto done;
+        }
+
+        for (i = 0; i < _ecall_table_size; i++)
+        {
+            if (strcmp(_ecall_table[i], name) == 0)
+            {
+                *global_id = i;
+                result = OE_OK;
+                break;
+            }
+        }
+
+        /* If the name is not found, adding it to the table. */
+        if (result == OE_NOT_FOUND)
+        {
+            if (_ecall_table_size >= OE_MAX_ECALLS)
+            {
+                result = OE_OUT_OF_BOUNDS;
+                goto done;
+            }
+
+            _ecall_table[_ecall_table_size] = name;
+            *global_id = _ecall_table_size;
+            _ecall_table_size++;
+            result = OE_OK;
+        }
+
+    done:
+        oe_lock.unlock();
+        return result;
+    }
+
+    static oe_result_t oe_get_enclave_function_id(
+        oe_enclave_t* enclave,
+        uint64_t* global_id,
+        const char* name,
+        uint64_t* id)
+    {
+        oe_result_t result = OE_FAILURE;
+
+        if (!enclave || !global_id || !name || !id)
+        {
+            result = OE_INVALID_PARAMETER;
+            goto done;
+        }
+
+        if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+        {
+            if ((result = oe_get_global_ecall_id_by_name(name, global_id)) !=
+                OE_OK)
+                goto done;
+        }
+
+        if (*global_id >= OE_MAX_ECALLS)
+        {
+            result = OE_OUT_OF_BOUNDS;
+            goto done;
+        }
+
+        /* Look-up the ecall id from the per-enclave table. */
+        *id = enclave->_ecall_info_table[*global_id].id;
+
+        result = OE_OK;
+    done:
+
+        return result;
+    }
+
+    static oe_result_t oe_host_register_enclave_functions(
```

Comment:
More comments on this: without passing the table a an argument of `oe_create_enclave`, we currently don't have good way for the host to reference the table defined in the `_u.c`.

Reply:
What I was thinking was something along the lines of this:

```c

// Have a type the contains the information about the ecall, and 
// which can be used for storing the global id as wll.
typedef struct _oe_ecall_info {
   const char* name;
   uint64_t global_id;
   // maybe in the future
   // uint64_t hash;
} oe_ecall_info_t;  // maybe choose a better name

typedef struct __oe_ecall_info_table {
    oe_ecall_info_t*  ecall_infos;
   uint32_t num_ecalls;
  uint64_t* cache;  // mapping from global id to local id
   uint64_t lock; // may for use as a spin lock
} oe_ecall_table_t; 

// Auto generate the info in _u.c:
static oe_ecall_info_t __ecall_infos[] = 
{
  { "ecall_1", OE_GLOBAL_ID_NULL}
   ...
};

static oe_ecall_table_t _ecall_table = 
{
   __ecall_infos, // pointer to infos array  above
   20, // number of ecalls
   NULL, // cache
   0  // lock initially zero.
};

// Then oe_host_register_enclave_functions could take in the oe_ecall_table_t instance.
// Which is called by oe_create_enclave.
static oe_result_t oe_host_register_enclave_functions(
        oe_enclave_t* enclave,
        oe_ecall_table_t* ecall_info_table);
// The above function could allocate and fill up the cache.
```

Reply:
How does host associate the `_ecall_table` with an enclave (with same or different binaries) inside a wrapper? It seems the same `_ecall_table` will be use in both scenarios.

Reply:
The _ecall table will be passed to oe_create_enclave instead of the names array. Alternatively,
we could generate a call to `oe_host_register_enclave_functions` in oe_create_<foo>_enclave function.

Reply:
Just to clarify, does this modification still keep the per-enlave table?
I thought this is for using the same ecall table for the instances of same enclave binary as you mention earlier and based on the comment:
`// Instead of each wrapper having a static global_id, it would just pass the array element of _ecall_table.`

Reply:
The ecall table and the global-id -> local-id cache within the ecall table structure will be shared by all enclave instances of the same enclave binary.

Reply:
Generating `oe_host_register_enclave_functions` may not work as there will be a conflict among multiple `_u.c` files.

I'm a bit unclear about how the host maintains the tables internally. Can you provide an example on how this works? Say the host creates three enclaves, two of them are the same binary.

Reply:
Here is one way to implement sharing of tables:
https://github.com/mingweishih/oeedger8r-cpp/compare/ecall_refactoring...anakrish:ak_ecall_refactoring_shared_table